content/general/concepts/abstract-factory/abstract-factory.md

```diff
@@ -0,0 +1,396 @@
+Description: 'The abstract factory pattern relies on a common interface for multiple factories and defers instantiation logic to their concrete model classes.'
+Subjects:
+  - 'Computer Science'
+  - 'Interview Prep'
+Tags:
+  - 'Classes'
+  - 'Conceptual'
+  - 'Constructors'
+  - 'Objects'
+CatalogContent:
+  - 'learn-java'
+  - 'paths/computer-science'
+---
+
+The abstract factory is often thought of as a factory of factory patterns. Similar to the factory pattern, the abstract factory relies on a common interface and defers instantiation logic to its concrete model classes. However, it covers multiple factories. Additionally, unlike the factory pattern, the abstract factory is typically built using composition whereas the factory is not.
+
+## UML Design
+
+![UML diagram of an abstract factory](https://raw.githubusercontent.com/Codecademy/docs/main/media/abstract-factory-uml.png)
+
+## Java Example
+
+To illustrate the abstract factory pattern, below is a real-world example, written in Java, that explores potential considerations for a banking account system. A new customer will request either a current or savings account. The customer may also be entitled to different privileges depending on their credit score. Below is a table of accounts and their privileges:
+
+| Customer Type | Current                     | Saving                |
+| ------------- | --------------------------- | --------------------- |
+| Gold          | Max overdraft limit of 3500 | Interest rate of 5%   |
+| Silver        | Max overdraft limit of 1200 | Interest rate of 3%   |
+| Bronze        | Max overdraft limit of 500  | Interest rate of 1.5% |
+| Builder       | No overdraft                | Interest rate of 1.5% |
+
+To simulate requesting and receiving a customer's credit score, a gateway has been mocked below. When given a customers name, the`CreditAgencyGateway` class should return an appropriate `Customer` object. We can later use this class to see the different paths through our abstract factory.
+
+```java
+public class CreditAgencyGateway {
+  // Depending on which name is searched for, a different customer object is returned
+  public Customer getCustomer(String name) {
+    return switch (name) {
+      case "Harry" -> new Customer("Harry",
+                                   "Potter",
+                                   LocalDate.of(1980, 7, 31),
+                                   "4 Privet Drive, Little Whinging",
+                                   795);
+      case "Ron" -> new Customer("Ron",
+                                 "Weasley",
+                                 LocalDate.of(1980, 3, 1),
+                                 "Shell Cottage, Cornwall",
+                                 379);
+      case "Hermione" -> new Customer("Hermione",
+                                      "Granger",
+                                      LocalDate.of(1979, 9, 19),
+                                      "12 Grimmauld Palace, London",
```

Comment:
```suggestion
                                      "Hampstead Garden Suburb, London",
```

12 Grimmauld is the house of the Blacks, the family of Harry's godfather Sirius 👀 🤣

Reply:
What a fun example ⚡

Reply:
Ye I think I must of del the wrong bit during one of the revisions. Sirius was one of the characters used in an old version of the gateway. He had a very bad credit rating due to his time in Azkaban 🤣